From 22898df6e0b1e15afe4bf2534bcd0231ffaaf8cf Mon Sep 17 00:00:00 2001
From: SimonFJ20 <simonfromjakobsen@gmail.com>
Date: Tue, 30 Nov 2021 01:22:17 +0100
Subject: [PATCH] added status code arg

---
 core/htb.py | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/core/htb.py b/core/htb.py
index b1944de..c864b9d 100644
--- a/core/htb.py
+++ b/core/htb.py
@@ -39,4 +39,5 @@ if __name__ == '__main__':
         if text.strip() == '':
             print(f"Empty file '{filepath}'")
             exit(1)
-        run_text(text, filepath, False)
+        print_res = '-s' in argv or '--status' in argv
+        run_text(text, filepath, print_res)
-- 
2.25.1

