From 6805d57096750ebc71492f441078c2eeb5461f95 Mon Sep 17 00:00:00 2001
From: Simon From <simonfromjakobsen@gmail.com>
Date: Wed, 24 Nov 2021 00:33:59 +0100
Subject: [PATCH] Initial commit

---
 README.md | 2 ++
 1 file changed, 2 insertions(+)
 create mode 100644 README.md

diff --git a/README.md b/README.md
new file mode 100644
index 0000000..dfb946b
--- /dev/null
+++ b/README.md
@@ -0,0 +1,2 @@
+# mamemalang
+Manual Memory Management Programming Language
-- 
2.25.1

