From 66a69a3d9a5e553dd5ecd790710c4ab114488d2b Mon Sep 17 00:00:00 2001
From: SimonFJ20 <simonfromjakobsen@gmail.com>
Date: Tue, 10 May 2022 02:09:08 +0200
Subject: [PATCH] added const to char*

---
 compiler.c | 4 ++--
 lexer.c    | 2 +-
 utils.c    | 2 +-
 utils.h    | 2 +-
 4 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/compiler.c b/compiler.c
index 20fc4a4..f50369e 100644
--- a/compiler.c
+++ b/compiler.c
@@ -60,7 +60,7 @@ void compiler_make_statement(Compiler* self, StatementNode* node)
 
 void compiler_make_function_definition(Compiler* self, FuncDefNode* node)
 {
-    char* name = chars_to_string((char*) node->target->value, node->target->length);
+    char* name = chars_to_string(node->target->value, node->target->length);
 
     self->inside_function = true;
     self->current_function_name = name;
@@ -99,7 +99,7 @@ void compiler_make_expression(Compiler* self, ExpressionNode* node)
 
 void compiler_make_int_literal(Compiler* self, IntNode* node)
 {
-    char* value_string = chars_to_string((char*) node->token->value, node->token->length);
+    char* value_string = chars_to_string(node->token->value, node->token->length);
     int value = atoi(value_string);
     free(value_string);
     string_builder_write_fmt(self->assembly, "    movl $%d, %%eax\n", value);
diff --git a/lexer.c b/lexer.c
index 3795678..53fbefe 100644
--- a/lexer.c
+++ b/lexer.c
@@ -92,7 +92,7 @@ void delete_token(Token* self)
 
 char* token_to_string(Token* self)
 {
-    char* value_str = chars_to_string((char*) self->value, self->length);
+    char* value_str = chars_to_string(self->value, self->length);
 
     char* buffer = calloc(8192, sizeof(char));
     snprintf(buffer, 8192, "Token(%s, '%s', %ld)", token_type_to_string(self->type), value_str, self->length);
diff --git a/utils.c b/utils.c
index 29e8e09..6e2e979 100644
--- a/utils.c
+++ b/utils.c
@@ -20,7 +20,7 @@ void println_and_free(char* string)
     free(string);
 }
 
-char* chars_to_string(char* chars, size_t amount)
+char* chars_to_string(const char* chars, size_t amount)
 {
     char* buffer = calloc(amount + 1, sizeof(char));
     strncpy(buffer, chars, amount);
diff --git a/utils.h b/utils.h
index 2c995bd..943ba93 100644
--- a/utils.h
+++ b/utils.h
@@ -117,7 +117,7 @@ char* string_builder_buffer(StringBuilder* self);
 void string_builder_write(StringBuilder* self, char* string);
 void string_builder_write_fmt(StringBuilder* self, const char* fmt, ...);
 
-char* chars_to_string(char* chars, size_t amount);
+char* chars_to_string(const char* chars, size_t amount);
 char* copy_string(const char* string);
 void println_and_free(char* string);
 void list_delete_all_and_self(List* list, void (*)(void*));
-- 
2.25.1

