From b9dd7b91b5daf1c2ef2fd27925a15c86972ecaaa Mon Sep 17 00:00:00 2001
From: SimonFJ20 <simonfromjakobsen@gmail.com>
Date: Thu, 5 May 2022 01:45:20 +0200
Subject: [PATCH] insure null in newly allocated bytes

---
 stringbuilder.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/stringbuilder.c b/stringbuilder.c
index 3e50a7b..5cb02aa 100644
--- a/stringbuilder.c
+++ b/stringbuilder.c
@@ -44,8 +44,10 @@ char* string_builder_buffer(StringBuilder* self)
 
 void string_builder_write(StringBuilder* self, char* string)
 {
+    size_t old_length = self->m_length;
     self->m_length += strlen(string);
     self->m_buffer = realloc(self->m_buffer, self->m_length * sizeof(char) + 1);
+    memset(self->m_buffer + sizeof(char) * old_length, '\0', self->m_length - old_length);
     strcat(self->m_buffer, string);
 }
 
-- 
2.25.1

